[WIP] Document for exposing and projecting binding metadata (#36310)
openshift/openshift-docs: modules/sbo-categories-of-exposable-binding-data.adoc

[id="sbo-categories-of-exposable-binding-data_{context}"]
= Categories of exposable binding data

> {servicebinding-title} provides the ability to expose the binding data values from the backing service resources and custom resource definitions (CRDs).

Reviewer: ```suggestion
The {servicebinding-title} enables you to expose the binding data values from the backing service resources and custom resource definitions (CRDs).
```
Re-word to make more user-centric

This section provides examples to show how you can use the various categories of exposable binding data.

== Exposing a string from a resource
.Example
[source,yaml]
----
apiVersion: postgres-operator.crunchydata.com/v1beta1
kind: PostgresCluster
metadata:
  name: hippo
  namespace: my-postgresql
  annotations:
    service.binding/username: path={.metadata.name}
    ...
----

> In the previous example the string from the `metadata.name` field of `PostgresCluster` custom resource (CR)  is exposed as a username.

Reviewer: ```suggestion
The following example shows how to expose the string from the `metadata.name` field of the `PostgresCluster` custom resource (CR) as a username:
```
Consider moving this sentence earlier to make it an intro to the code sample so that the user knows what they're seeing in the sample. 

If you move it, edit appropriately to make it an introductory sentence and to avoid the passive voice, per my suggestions.

== Exposing an entire config map or secret that is referenced from a resource

> Following are the examples for exposing an entire secret as annotations:

Reviewer: ```suggestion
The following examples show how to expose an entire secret as annotations:
```
Edit to make the sentence shorter and more active.

.Example: Exposing an entire secret as annotations
[source,yaml]
----
apiVersion: postgres-operator.crunchydata.com/v1beta1
kind: PostgresCluster
metadata:
  name: hippo
  namespace: my-postgresql
  annotations:

>     dbsecret: hippo-pguser-hippo
>     service.binding: 'path={.metadata.annotations.dbsecret},objectType=Secret'

Reviewer 2: ```suggestion
    service.binding: 'path={.metadata.name}-pguser-{.metadata.name},objectType=Secret'
```

----

.Example: The referenced secret from the backing service resource
[source,yaml]
----
apiVersion: v1
kind: Secret
metadata:
  name: hippo-pguser-hippo
data:
  password: "MTBz"
  user: "Z3Vlc3Q="
----

> Following is the example for exposing an entire config map as OLM descriptors:

Reviewer: ```suggestion
The following example shows how to expose an entire config map as OLM descriptors:
```
Edit to make the sentence shorter and more active.

.Example: Exposing an entire config map as OLM descriptors
[source,yaml]
----
- path: data.dbConfiguration
  x-descriptors:
  - urn:alm:descriptor:io.kubernetes:ConfigMap
  - service.binding
----

The previous example uses the `path` attribute with a `urn:alm:descriptor:io.kubernetes:ConfigMap` entry to indicate that the path points to the `ConfigMap` service resource.

If you intend to project all the values from a `ConfigMap` service resource, you must specify it as an attribute in the backing service CR. For example, if the attribute is part of the `.spec` section, you can create and use a `specDescriptors` array. Or, if the attribute is part of the `.status` section, you can create and use a `statusDescriptors` array.

== Exposing a specific entry from a config map or secret that is referenced from a resource

> Following are the examples for exposing a specific entry from a config map as annotations:

Reviewer: ```suggestion
The following examples show how to expose a specific entry from a config map as annotations:
```
Edit to make the sentence more active and direct.

.Example: Exposing an entry from a config map as annotations
[source,yaml]
----
apiVersion: postgres-operator.crunchydata.com/v1beta1
kind: PostgresCluster
metadata:
  name: hippo
  namespace: my-postgresql
  annotations:
    dbconfig: hippo-config
    service.binding: 'path={.metadata.annotations.dbconfig},objectType=ConfigMap,sourceKey=user'
----

.Example: The referenced config map from the backing service resource
The binding data should have a key with name as `timeout` and value as `10s`:
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: hippo-config
data:
  db_timeout: "10s"
  user: "hippo"
----

> Following is the example for exposing a specific entry from a config map as OLM descriptors:

Reviewer: ```suggestion
The following example shows how to expose a specific entry from a config map as OLM descriptors:
```
Edit to make the sentence more active and direct

.Example: Exposing an entry from a config map as OLM descriptors
[source,yaml]
----
- path: data.dbConfiguration
  x-descriptors:
  - urn:alm:descriptor:io.kubernetes:ConfigMap
  - service.binding:my_certificate:sourceKey=certificate
----

The previous example uses the `path` attribute with an `X-Descriptors` update for `service.binding` and `sourceKey` by providing the following information:

* Name of the binding key that is to be injected
* Name of the key in the Secret service resource

== Exposing a resource definition value

> Following is the example for exposing a resource definition value as annotations:

Reviewer: ```suggestion
The following example shows how to expose a resource definition value as annotations:
```
Edit to make the sentence more active and direct
